src/ImageSharp/Formats/Jpg/JpegDecoderCore.cs

```diff
@@ -1337,902 +766,688 @@ private void ConvertFromYCbCr<TColor>(int width, int height, Image<TColor> image
                 }
 
-                if (ah != 0 && ah != al + 1)
+                int th = this.Temp[0] & 0x0f;
+                if (th > HuffmanTree.MaxTh || (!this.IsProgressive && (th > 1)))
                 {
-                    throw new ImageFormatException("Bad successive approximation values");
+                    throw new ImageFormatException("Bad Th value");
                 }
-            }
-
-            // mxx and myy are the number of MCUs (Minimum Coded Units) in the image.
-            int h0 = this.componentArray[0].HorizontalFactor;
-            int v0 = this.componentArray[0].VerticalFactor;
-            int mxx = (this.imageWidth + (8 * h0) - 1) / (8 * h0);
-            int myy = (this.imageHeight + (8 * v0) - 1) / (8 * v0);
 
-            if (this.grayImage == null && this.ycbcrImage == null)
-            {
-                this.MakeImage(mxx, myy);
+                int huffTreeIndex = (tc * HuffmanTree.ThRowSize) + th;
+                this.HuffmanTrees[huffTreeIndex].ProcessDefineHuffmanTablesMarkerLoop(this, this.Temp, ref remaining);
             }
+        }
 
-            if (this.isProgressive)
+        /// <summary>
+        /// Processes the DRI (Define Restart Interval Marker) Which specifies the interval between RSTn markers, in
+        /// macroblocks
+        /// </summary>
+        /// <param name="remaining">The remaining bytes in the segment block.</param>
+        private void ProcessDefineRestartIntervalMarker(int remaining)
+        {
+            if (remaining != 2)
             {
-                for (int i = 0; i < scanComponentCount; i++)
-                {
-                    int compIndex = scan[i].Index;
-                    if (this.progCoeffs[compIndex] == null)
-                    {
-                        int size = mxx * myy * this.componentArray[compIndex].HorizontalFactor
-                                   * this.componentArray[compIndex].VerticalFactor;
-
-                        this.progCoeffs[compIndex] = new Block8x8F[size];
-                    }
-                }
+                throw new ImageFormatException("DRI has wrong length");
             }
 
-            this.bits = default(Bits);
-
-            int mcu = 0;
-            byte expectedRst = JpegConstants.Markers.RST0;
-
-            // b is the decoded coefficients block, in natural (not zig-zag) order.
-            // Block b;
-            int[] dc = new int[MaxComponents];
-
-            // bx and by are the location of the current block, in units of 8x8
-            // blocks: the third block in the first row has (bx, by) = (2, 0).
-            int bx, by, blockCount = 0;
-
-            Block8x8F b = default(Block8x8F);
-            Block8x8F temp1 = default(Block8x8F);
-            Block8x8F temp2 = default(Block8x8F);
-
-            UnzigData unzig = UnzigData.Create();
-
-            int* unzigPtr = unzig.Data;
+            this.ReadFull(this.Temp, 0, 2);
```

Comment:
The `remaining` parameter is only used for precondition checks. Really this line should be. `this.ReadFull(this.Temp, 0, remaining); `